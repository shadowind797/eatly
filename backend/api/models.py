from django.db import models
from django.contrib.auth.models import AbstractUser


class User(AbstractUser):
    username = models.CharField(max_length=120, unique=True)
    password = models.CharField(max_length=120)
    phone = models.CharField(max_length=120, blank=True)
    status = models.ForeignKey('Status', on_delete=models.PROTECT, default=5)
    is_banned = models.BooleanField(default=False)
    ban_reason = models.TextField(blank=True)
    passwd_change_link = models.CharField(max_length=120, blank=True)

    def __str__(self):
        return self.username


class Item(models.Model):
    title = models.CharField(max_length=120, unique=True)
    description = models.TextField()
    price = models.FloatField()
    photo = models.ImageField(upload_to='img')
    category = models.ForeignKey('Category', on_delete=models.SET_NULL, null=True, blank=True)
    restaurant = models.ForeignKey('Restaurant', on_delete=models.SET_NULL, null=True, blank=True)
    rating = models.FloatField()

    def __str__(self):
        return self.title


class Restaurant(models.Model):
    name = models.CharField(max_length=120, unique=True)
    address = models.CharField(max_length=120)
    rating = models.FloatField()
    image = models.ImageField(upload_to='img/rests', null=True)
    category = models.ForeignKey('RestaurantCat', on_delete=models.SET_NULL, null=True, blank=True)

    def __str__(self):
        return self.name


class Address(models.Model):
    house_address = models.CharField(max_length=120)
    entrance = models.CharField(max_length=120)
    floor = models.CharField(max_length=120)
    flat = models.CharField(max_length=120)
    owner = models.ForeignKey('User', on_delete=models.CASCADE)
    active = models.BooleanField(default=False)

    def __str__(self):
        return self.house_address


class Order(models.Model):
    user = models.ForeignKey('User', on_delete=models.SET_NULL, null=True)
    payment = models.ForeignKey('Payments', on_delete=models.SET_NULL, null=True)
    address = models.ForeignKey('Address', on_delete=models.SET_NULL, null=True)
    total = models.FloatField()
    status = models.ForeignKey('OrderStatus', on_delete=models.SET_NULL, null=True)
    restaurant = models.ForeignKey('Restaurant', on_delete=models.SET_NULL, null=True)
    coupon = models.ForeignKey('Coupon', on_delete=models.SET_NULL, null=True)
    comment = models.TextField(null=True, blank=True)

    def __str__(self):
        return self.user


class OrderStatus(models.Model):
    name = models.CharField(max_length=120, unique=True)


class Coupon(models.Model):
    title = models.CharField(max_length=120, unique=True)
    valid_from = models.DateField(auto_now_add=True)
    valid_to = models.DateField()
    is_valid = models.BooleanField(default=True)
    value = models.FloatField()
    times_activated = models.IntegerField(default=0)
    enabled_times = models.IntegerField()
    category = models.ForeignKey('CouponCategory', on_delete=models.CASCADE)

    def __str__(self):
        return self.title


class CouponCategory(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Payments(models.Model):
    number = models.CharField(max_length=120, unique=True)
    date_to = models.CharField(max_length=100)
    cvv = models.IntegerField()
    owner = models.ForeignKey('User', on_delete=models.CASCADE)
    active = models.BooleanField(default=False)
    name = models.CharField(max_length=120, default="")

    def __str__(self):
        return self.number


class CartItem(models.Model):
    item = models.ForeignKey('Item', on_delete=models.CASCADE)
    quantity = models.IntegerField(default=1)
    owner = models.ForeignKey('User', on_delete=models.CASCADE)

    def __str__(self):
        return self.item.title


class Reviews(models.Model):
    author = models.ForeignKey('User', on_delete=models.CASCADE)
    time = models.CharField(max_length=120)
    text = models.TextField()
    rating = models.FloatField()
    created_at = models.DateTimeField(auto_now_add=True)

    def __str__(self):
        return self.author.username


class RestaurantCat(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Category(models.Model):
    name = models.CharField(max_length=120, unique=True)

    def __str__(self):
        return self.name


class Status(models.Model):
    name = models.CharField(max_length=120)

    def __str__(self):
        return self.name
