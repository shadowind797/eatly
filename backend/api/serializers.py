from .models import *
from rest_framework import serializers


class UserSerializer(serializers.ModelSerializer):
    class Meta:
        model = User
        fields = ['id', 'username', 'email', 'password', "phone", "status", "is_banned", "ban_reason", "first_name"]
        extra_kwargs = {"password": {"write_only": True}, "status": {"read_only": True}}

    def create(self, validated_data):
        user = User.objects.create_user(**validated_data)
        return user


class AddressSerializer(serializers.ModelSerializer):
    class Meta:
        model = Address
        fields = ["id", "house_address", "entrance", "floor", "flat"]

    def create(self, validated_data):
        address = Address.objects.create_address(**validated_data)
        return address


class PaymentSerializer(serializers.ModelSerializer):
    class Meta:
        model = Payments
        fields = "__all__"
        extra_kwargs = {"cvv": {"write_only": True}, "date_to": {"read_only": True}}

    def create(self, validated_data):
        payment = Payments.objects.create_payments(**validated_data)
        return payment


class OrderSerializer(serializers.ModelSerializer):
    class Meta:
        model = Order
        fields = ["id", "status", "total", "payment", "address"]

    def create(self, validated_data):
        order = Order.objects.create_order(**validated_data)
        return order


class ItemSerializer(serializers.ModelSerializer):
    class Meta:
        model = Item
        fields = ['id', "title", "description", "price", "photo", "category", "rating"]


class CartItemSerializer(serializers.ModelSerializer):
    class Meta:
        model = CartItem
        fields = ['id', "item", "quantity"]


class CategorySerializer(serializers.ModelSerializer):
    class Meta:
        model = Category
        fields = ['id', "name"]


class RestaurantSerializer(serializers.ModelSerializer):
    class Meta:
        model = Restaurant
        fields = ['id', "name", "address", "rating", "category_id", "image"]


class RestaurantCatSerializer(serializers.ModelSerializer):
    class Meta:
        model = RestaurantCat
        fields = ['id', "name"]
