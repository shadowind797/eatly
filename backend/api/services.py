from django.conf import settings
from django.shortcuts import redirect
from django.core.exceptions import ValidationError
from urllib.parse import urlencode
from typing import Dict, Any
import requests
from .models import User
import jwt

GOOGLE_ACCESS_TOKEN_OBTAIN_URL = 'https://oauth2.googleapis.com/token'
GOOGLE_USER_INFO_URL = 'https://www.googleapis.com/oauth2/v3/userinfo'

GITHUB_ACCESS_TOKEN_OBTAIN_URL = 'https://github.com/login/oauth/access_token'
GITHUB_USER_INFO_URL = 'https://api.github.com/user'

LOGIN_URL = f'{settings.BASE_APP_URL}/login'


def google_get_access_token(code: str, redirect_uri: str) -> str:
    data = {
        'code': code,
        'client_id': settings.GOOGLE_OAUTH2_CLIENT_ID,
        'client_secret': settings.GOOGLE_OAUTH2_CLIENT_SECRET,
        'redirect_uri': redirect_uri,
        'grant_type': 'authorization_code'
    }

    response = requests.post(GOOGLE_ACCESS_TOKEN_OBTAIN_URL, data=data)
    if not response.ok:
        raise ValidationError('Could not get access token from Google.')

    access_token = response.json()['access_token']

    return access_token


def google_get_user_info(access_token: str) -> Dict[str, Any]:
    response = requests.get(
        GOOGLE_USER_INFO_URL,
        params={'access_token': access_token}
    )

    if not response.ok:
        raise ValidationError('Could not get user info from Google.')

    return response.json()


def github_get_access_token(code: str, redirect_uri: str) -> str:
    headers = {
        "Accept": "application/json"
    }
    data = {
        'code': code,
        'client_id': settings.GITHUB_OAUTH2_CLIENT_ID,
        'client_secret': settings.GITHUB_OAUTH2_CLIENT_SECRET,
        'redirect_uri': redirect_uri
    }

    response = requests.post(GITHUB_ACCESS_TOKEN_OBTAIN_URL, headers=headers, data=data)
    if not response.ok:
        raise ValidationError('Could not get access token from Github.')

    access_token = response.json()['access_token']

    return access_token


def github_get_user_info(access_token: str) -> Dict[str, Any]:
    headers = {
        'Authorization': f'Bearer {access_token}'
    }
    response = requests.get(
        GITHUB_USER_INFO_URL,
        headers=headers
    )

    if not response.ok:
        raise ValidationError('Could not get user info from Github.')

    return response.json()


def get_user_data(validated_data, platform):
    domain = settings.BASE_API_URL
    redirect_uri = f'{domain}/api/login/{platform}/'

    code = validated_data.get('code')
    error = validated_data.get('error')
    state = validated_data.get('state')

    if error or not code:
        params = urlencode({'error': error})
        return redirect(f'{LOGIN_URL}?{params}')

    if platform == "google":
        access_token = google_get_access_token(code=code, redirect_uri=redirect_uri)
        user_data = google_get_user_info(access_token=access_token)

        user, created = User.objects.get_or_create(
            email=user_data['email'],
        )

        if not created:
            user.username = user_data['name']
            user.save()

        profile_data = {
            'email': user_data['email'],
        }
        return profile_data
    elif platform == "github":
        access_token = github_get_access_token(code=code, redirect_uri=redirect_uri)
        user_data = github_get_user_info(access_token=access_token)

        print(user_data)

        User.objects.get_or_create(
            username=user_data['login'],
        )

        profile_data = {
            'username': user_data['login'],
        }
        return profile_data
