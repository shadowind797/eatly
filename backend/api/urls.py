from django.urls import path, include

from . import views

urlpatterns = [
    path('access/', views.GetAccess.as_view(), name='access'),
    path('user/', views.GetUser.as_view(), name='user'),
    path('user/change/', views.GetUser.as_view(), name='user-change'),

    path("login/google/", views.GoogleLoginApi.as_view(), name='google-login'),
    path("login/github/", views.GithubLoginApi.as_view(), name='github-login'),

    path('items/', views.ItemListCreate.as_view(), name='item-list'),
    path('items/top/', views.ItemListCreate.as_view(), name='item-list'),
    path('items/category/', views.CategoriesList.as_view(), name='item-category-list'),

    path('items/cart/', views.CartItemListCreate.as_view(), name='item-cart-list'),
    path('items/cart/add', views.CartItemListCreate.as_view(), name='item-cart-list'),
    path('items/cart/delete', views.DeleteCartItem.as_view(), name='item-cart-list'),
    path('items/cart/check', views.CheckInCart.as_view(), name='item-cart-list'),

    path("items/delete/<int:pk>/", views.ItemDelete.as_view(), name='item-delete'),

    path("restaurants/", views.RestaurantListCreate.as_view(), name='restaurant-list'),
    path("restaurants/categories/", views.RestaurantCategories.as_view(), name='restaurant-cats-list'),

    path("address/", views.AddressList.as_view(), name='address-list'),
    path("address/add/", views.AddressList.as_view(), name='restaurant-cats-list'),

    path("order/", views.OrderView.as_view(), name='order'),
    path("order/add/", views.OrderView.as_view(), name='order-create'),
    path("order/cancel/", views.CancelOrder.as_view(), name='order-cancel'),

    path("payment/", views.PaymentView.as_view(), name='payment'),
    path("payment/add/", views.PaymentView.as_view(), name='payment-create'),

    path("coupon/", views.CouponListCreate.as_view(), name='coupon-list'),
    path("coupon/add/", views.CouponListCreate.as_view(), name='coupon-create'),

    path("items/search/", views.SearchView.as_view(), name='search'),
    path("items/search/filters/", views.FilterView.as_view(), name='filter'),

    path("profile/orders", views.GetProfile.as_view(), name='all-user-orders'),
]
