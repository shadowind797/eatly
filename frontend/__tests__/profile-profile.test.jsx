import { render, fireEvent, waitFor, act } from "@testing-library/react";
import api from "../src/api.js";
import { MemoryRouter as Router } from "react-router-dom";
import ProfileContentProfile from "../src/components/Profile/PContent";

jest.mock("../src/api.js");

describe("Profile component -> Profile", () => {
  const user = {
    id: 1,
    username: "shadowind",
    email: "shadowind797@gmail.com",
    phone: "+375666999",
    status: 1,
    is_banned: false,
    ban_reason: "",
    first_name: "Peter",
  };

  const orders = {
    orders: [
      {
        id: 46,
        status: 2,
        total: 64.86,
        payment: null,
        address: 11,
        created: "04.09 09:19",
        rest: "Mc'Donalds",
        items_count: 4,
      },
      {
        id: 45,
        status: 2,
        total: 24.19,
        payment: 2,
        address: 10,
        created: "03.09 11:50",
        rest: "Evos",
        items_count: 1,
      },
      {
        id: 44,
        status: 2,
        total: 20.89,
        payment: null,
        address: 13,
        created: "03.09 11:46",
        rest: "#Farsh",
        items_count: 1,
      },
    ],
    statuses: [
      {
        id: 5,
        name: "Canceled",
      },
      {
        id: 4,
        name: "Completed",
      },
      {
        id: 3,
        name: "Delivery",
      },
      {
        id: 2,
        name: "In progress",
      },
      {
        id: 1,
        name: "Staged",
      },
    ],
  };

  beforeEach(() => {
    api.post.mockImplementation((url) => {
      if (url.includes("api/password/change/")) {
        return Promise.resolve({
          status: 200,
        });
      }
      return Promise.reject(new Error("Not found"));
    });
  });

  it("renders without crashing", () => {
    render(
      <Router>
        <ProfileContentProfile
          user={user}
          orders={orders.orders}
          osl={orders.statuses}
        />
      </Router>
    );
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(
      <Router>
        <ProfileContentProfile
          user={user}
          orders={orders.orders}
          osl={orders.statuses}
        />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("sends email to change password", async () => {
    const { getByText } = render(
      <Router>
        <ProfileContentProfile
          user={user}
          orders={orders.orders}
          osl={orders.statuses}
        />
      </Router>
    );

    await act(async () => {
      fireEvent.click(getByText("Change password"));
      waitFor(() => getByText("Send email"));
    });

    await act(async () => {
      fireEvent.click(getByText("Send email"));
    });

    expect(api.post).toHaveBeenCalledWith("api/password/change/", {
      method: "send_email",
    });
  });
});
