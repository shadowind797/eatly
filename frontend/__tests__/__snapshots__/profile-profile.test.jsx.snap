// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component -> Profile matches the snapshot 1`] = `
<DocumentFragment>
  <div
    id="profile-content-profile"
  >
    <div
      id="first"
    >
      <div
        id="profile"
      >
        <div
          id="profile-info"
        >
          <div
            id="photo-name"
          >
            <img
              alt=""
              id="user-img"
              src="test-file-stub"
            />
            <div
              id="top-info"
            >
              <h3>
                Username: 
                <span>
                  shadowind
                </span>
              </h3>
              <p>
                Name: 
                <span>
                  Peter
                </span>
              </p>
            </div>
          </div>
          <div
            id="others"
          >
            <div>
              <img
                alt=""
                src="test-file-stub"
              />
              <p>
                ID: 
                <span>
                  1
                </span>
              </p>
            </div>
            <div>
              <img
                alt=""
                src="test-file-stub"
              />
              <p>
                Email: 
                <span>
                  shadowind797@gmail.com
                </span>
              </p>
            </div>
            <div>
              <img
                alt=""
                src="test-file-stub"
              />
              <p>
                Phone: 
                <span>
                  +375666999
                </span>
              </p>
            </div>
          </div>
        </div>
        <div
          id="profile-actions"
        >
          <button
            id="change-pass"
          >
            <img
              alt="Change password"
              src="test-file-stub"
            />
            <p>
              Change password
            </p>
          </button>
          <button
            id="logout"
          >
            <img
              alt="Log out"
              src="test-file-stub"
            />
            <p>
              Log out
            </p>
          </button>
          <button
            class="disabled"
            id="delete-acc"
          >
            <img
              alt="Delete account"
              src="test-file-stub"
            />
            <p>
              Delete account
            </p>
          </button>
        </div>
      </div>
      <h2>
        Your latest orders
      </h2>
      <div
        id="last-orders"
      >
        <div
          id="order-card"
        >
          <div
            class="head"
          >
            <p>
              04.09 09:19
            </p>
            <p
              class="last"
            >
              In progress
            </p>
          </div>
          <div
            class="info"
          >
            <div
              id="rest"
            >
              <img
                alt=""
                src=""
              />
              <h3>
                Mc'Donalds
              </h3>
            </div>
            <h3>
              4 dishes
            </h3>
            <h3
              class="order-total"
            >
              $64.86
            </h3>
          </div>
        </div>
        <div
          id="order-card"
        >
          <div
            class="head"
          >
            <p>
              03.09 11:50
            </p>
            <p
              class="last"
            >
              In progress
            </p>
          </div>
          <div
            class="info"
          >
            <div
              id="rest"
            >
              <img
                alt=""
                src=""
              />
              <h3>
                Evos
              </h3>
            </div>
            <h3>
              1 dishes
            </h3>
            <h3
              class="order-total"
            >
              $24.19
            </h3>
          </div>
        </div>
        <div
          id="order-card"
        >
          <div
            class="head"
          >
            <p>
              03.09 11:46
            </p>
            <p
              class="last"
            >
              In progress
            </p>
          </div>
          <div
            class="info"
          >
            <div
              id="rest"
            >
              <img
                alt=""
                src=""
              />
              <h3>
                #Farsh
              </h3>
            </div>
            <h3>
              1 dishes
            </h3>
            <h3
              class="order-total"
            >
              $20.89
            </h3>
          </div>
        </div>
      </div>
    </div>
    <div
      id="statistics"
    >
      <div>
        <h2>
          Unlock spending statistics
        </h2>
        <h3>
          <span
            id="wt"
          >
            with
          </span>
           EATLY 
          <span
            id="prem"
          >
            Premium
          </span>
        </h3>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
